Mathlib/RingTheory/Flat/Basic.lean

```diff
@@ -64,268 +65,293 @@ open Function (Surjective)
-
-instance {S} [CommRing S] [Algebra R S] [Module S M] [IsScalarTower R S M] [Flat S M] [Flat R N] :
-    Flat S (M ⊗[R] N) :=
-  (iff_rTensor_injective' _ _).mpr fun I ↦ by
+instance of_projective [Projective R M] : Flat R M :=
+  have ⟨e, he⟩:= Module.projective_def'.mp ‹_›
+  of_retract _ _ he
+
+@[deprecated (since := "2024-12-26")] alias of_projective_surjective := of_projective
+
+instance of_free [Free R M] : Flat R M := inferInstance
+
+instance {S} [CommSemiring S] [Algebra R S] [Module S M] [IsScalarTower R S M]
+    [Flat S M] [Flat R N] : Flat S (M ⊗[R] N) :=
+  iff_rTensor_injectiveₛ.mpr fun P _ _ I ↦ by
+    letI := RestrictScalars.moduleOrig R S P
+    change Submodule S (RestrictScalars R S P) at I
+    change Function.Injective (rTensor _ I.subtype)
     simpa [AlgebraTensorModule.rTensor_tensor] using
       rTensor_preserves_injective_linearMap (.restrictScalars R <| I.subtype.rTensor M)
       (rTensor_preserves_injective_linearMap _ I.injective_subtype)
 
 example [Flat R M] [Flat R N] : Flat R (M ⊗[R] N) := inferInstance
 
-/--
-If `M` is a flat module, then `𝟙 M ⊗ f` is injective for all injective linear maps `f`.
--/
-theorem lTensor_preserves_injective_linearMap {N' : Type*} [AddCommGroup N'] [Module R N']
-    [Flat R M] (L : N →ₗ[R] N') (hL : Function.Injective L) :
-    Function.Injective (L.lTensor M) :=
-  (L.lTensor_inj_iff_rTensor_inj M).2 (rTensor_preserves_injective_linearMap L hL)
-
-theorem linearIndependent_one_tmul {S} [Ring S] [Algebra R S] [Flat R S] {ι} {v : ι → M}
+theorem linearIndependent_one_tmul {S} [Semiring S] [Algebra R S] [Flat R S] {ι} {v : ι → M}
     (hv : LinearIndependent R v) : LinearIndependent S ((1 : S) ⊗ₜ[R] v ·) := by
   classical rw [LinearIndependent, ← LinearMap.coe_restrictScalars R,
     Finsupp.linearCombination_one_tmul]
   simpa using lTensor_preserves_injective_linearMap _ hv
 
-variable (R M) in
+variable (p : Submodule R M) (q : Submodule R N)
+
+/-- If p and q are submodules of M and N respectively, and M and q are flat,
+then `p ⊗ q → M ⊗ N` is injective. -/
+theorem tensorProduct_mapIncl_injective_of_right
+    [Flat R M] [Flat R q] : Function.Injective (mapIncl p q) := by
+  rw [mapIncl, ← lTensor_comp_rTensor]
+  exact (lTensor_preserves_injective_linearMap _ q.injective_subtype).comp
+    (rTensor_preserves_injective_linearMap _ p.injective_subtype)
+
+/-- If p and q are submodules of M and N respectively, and N and p are flat,
+then `p ⊗ q → M ⊗ N` is injective. -/
+theorem tensorProduct_mapIncl_injective_of_left
+    [Flat R p] [Flat R N] : Function.Injective (mapIncl p q) := by
+  rw [mapIncl, ← rTensor_comp_lTensor]
+  exact (rTensor_preserves_injective_linearMap _ p.injective_subtype).comp
+    (lTensor_preserves_injective_linearMap _ q.injective_subtype)
+
+/-! ### Flatness over a ring -/
+
+variable {R : Type u} {M : Type v} [CommRing R] [AddCommGroup M] [Module R M]
+variable {N : Type w} [AddCommGroup N] [Module R N]
```

Comment:
I think it would be good to wrap the semiring parts in a section, and start a new section here.